### DISCUSSION TITLE
Download from Shared Asset library

### DISCUSSION THREAD

**Discussion by gammelgaard52** [d365collaborative/d365fo.tools#505] [Q&A] [answered]
Is it possible to implement some logic, to download VHD files from LCS Asset Library? And I guess it depend a lot on Microsoft, if the API is available..

**Comment by Splaxi** [comment, marked as answer]
You will need to hunt down either Lane or Billy from Microsoft, so they learn that there's a need for this. 

But to be fair, it takes 2 click per file, to obtain a full uri with SAS token, that allows you to leverage the double hop storage account transfer, found on the wiki page.

This will take your download time from hours per file, to minutes.

https://github.com/d365collaborative/d365fo.tools/wiki/How-To-Transfer-Via-AzCopy

**Reply by Splaxi** [reply]
The wiki is a bit outdated on the part that will get you the uri, but the rest just works.

**Reply by gammelgaard52** [reply]
The reason I ask, is because of a very inspiring session at FinOps Bootcamp today, where they demonstrated another way to create build VMs for your DevOps build pipelines. D365FO VHD VM were part of it and to be entirely automated, that download should also be handled automatically :)

Thx for the answer Mötz :)

**Comment by Splaxi** [comment]
If you have them stored in your own storage account, then you should be able to run the same scenario as they showed.

**Reply by Alpa-Chino** [reply]
@gammelgaard52 @Splaxi : sorry to barge in, what kind of way is that you are mentioning ? Is there a doc or  video?

**Reply by Splaxi** [reply]
The video wasn't recorded. But I think you should see the video from Joris and the dynamics con from last year. It sparked the idea from Adria and Juan, which are the Spanish guys that ran the session we are talking about. 

I'll expect them to blog about it over the next few weeks.

**Reply by Alpa-Chino** [reply]
Thanks Mötz! I will watch it and i'll be on the lookout for their blogpost!

**Reply by Splaxi** [reply]
Rumor has it that they in fact have the session recorded by themselves and they will be posting it in the near future. 

The tools only play an initial part and that's to speed up the download of the vhd files and nothing more. But I'll let you know when I know more and see if we can fill out the gaps.

**Reply by gammelgaard52** [reply]
It's really interesting and I think there is a lot of potential for the broader Dynamics world, to do more automation, with the purpose of streamlining internal processes and move the more technical tasks, to a more user friendly flow.

**Comment by smholvoet** [comment]
Related: is it possible to target the **shared** asset library when using the `Get-D365LcsAssetFile`? Judging by the above replies, I guess not?

Use case 👉 fetching the NuGet packages of the latest service update:

```powershell
Get-D365LcsAssetFile -FileType "NuGetPackage" -AssetName "*10.0.25*"
```

**Reply by FH-Inway** [reply]
Agreed, that would be nice. However, as far as I know, the LCS API does not allow access to the shared asset library, only to the project asset libraries. Then again, you only need to import the NuGet packages into a project asset library manually. After that, you can use `Get-D365LcsAssetFile`. So, 8 times a year?

**Reply by Splaxi** [reply]
As mentioned - the **Shared Asset Library** is off limits from the API perspective.

You can import the files from the Shared Asset Library and then you will be able to download them via the API. At best this is an incomplete story from Microsoft, at I don't see they are going to complete it in the near future.

**Reply by smholvoet** [reply]
Alright, I'll be importing the packages from the shared asset library to the private one as a workaround. Adds one extra step but beats doing everything manually.

**Reply by FH-Inway** [reply]
This is now possible with the [Get-D365LcsSharedAssetFile](https://github.com/d365collaborative/d365fo.tools/wiki/Get-D365LcsSharedAssetFile) cmdlet.
